src/dbnode/storage/series/buffer.go

```diff
@@ -181,287 +161,281 @@ func (b *dbBuffer) Write(
 	}
-}
 
-// computedForEachBucketAsc performs a fn on the buckets in time ascending order
-// and returns the sum of the number returned by each fn
-func (b *dbBuffer) computedForEachBucketAsc(
-	op computeBucketIdxOp,
-	fn func(now time.Time, b *dbBuffer, idx int, bucketStart time.Time) int,
-) int {
-	now := b.nowFn()
-	pastMostBucketStart := now.Truncate(b.blockSize).Add(-1 * b.blockSize)
-	bucketNum := (pastMostBucketStart.UnixNano() / int64(b.blockSize)) % bucketsLen
-	result := 0
-	for i := int64(0); i < bucketsLen; i++ {
-		idx := int((bucketNum + i) % bucketsLen)
-		curr := pastMostBucketStart.Add(time.Duration(i) * b.blockSize)
-		result += fn(now, b, idx, curr)
+	// A call to flush can only be for warm writes
+	block, err := buckets.toBlock(WarmWrite)
+	if err != nil {
+		return FlushOutcomeErr, err
 	}
-	if op == computeAndResetBucketIdx {
-		b.pastMostBucketIdx = int(bucketNum)
+	if block == nil {
+		return FlushOutcomeBlockDoesNotExist, nil
 	}
-	return result
-}
 
-func (b *dbBuffer) Snapshot(ctx context.Context, blockStart time.Time) (xio.SegmentReader, error) {
-	var (
-		res xio.SegmentReader
-		err error
-	)
-
-	b.forEachBucketAsc(func(bucket *dbBufferBucket) {
-		if err != nil {
-			// Something already went wrong and we want to return the error to the caller
-			// as soon as possible instead of continuing to do work.
-			return
-		}
-
-		if !bucket.canRead() {
-			return
-		}
+	stream, err := block.Stream(ctx)
+	if err != nil {
+		return FlushOutcomeErr, err
+	}
 
-		if !blockStart.Equal(bucket.start) {
-			return
-		}
+	segment, err := stream.Segment()
+	if err != nil {
+		return FlushOutcomeErr, err
+	}
 
-		// We need to merge all the bootstrapped blocks / encoders into a single stream for
-		// the sake of being able to persist it to disk as a single encoded stream.
-		_, err = bucket.merge()
-		if err != nil {
-			return
-		}
+	checksum, err := block.Checksum()
+	if err != nil {
+		return FlushOutcomeErr, err
+	}
 
-		// This operation is safe because all of the underlying resources will respect the
-		// lifecycle of the context in one way or another. The "bootstrapped blocks" that
-		// we stream from will mark their internal context as dependent on that of the passed
-		// context, and the Encoder's that we stream from actually perform a data copy and
-		// don't share a reference.
-		streams := bucket.streams(ctx)
-		if len(streams) != 1 {
-			// Should never happen as the call to merge above should result in only a single
-			// stream being present.
-			err = errMoreThanOneStreamAfterMerge
-			return
-		}
+	err = persistFn(id, tags, segment, checksum)
+	if err != nil {
+		return FlushOutcomeErr, err
+	}
 
-		// Direct indexing is safe because canRead guarantees us at least one stream
-		res = streams[0]
-	})
+	// Don't need to check error here because a non-nil block means that a
+	// writable bucket exists
+	bucket, _ := buckets.writableBucket(WarmWrite)
+	bucket.version = version
```

Comment:
probably need to set the version on the bootstrapped blocks too so we can evict them in the next tick

Reply:
When bootstrapping a block, it actually goes into the writable bucket (the same bucket where writes go to). Maybe I should rename writable bucket to mutable bucket to be clearer.